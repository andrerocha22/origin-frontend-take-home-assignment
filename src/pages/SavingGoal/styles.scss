@import '~/assets/styles/global.scss';

.savingGoalSimulator {
  max-width: 42vw;
  min-width: 42vw;

  & h3 {
    text-align: center;
    padding: 3.5vh 0 0 0;
    color: $blue-1;
    font-weight: normal;
    font-size: calc(#{$font} * 2);
    line-height: 32px;
  }

  & .goalCalculator {
    background-color: $white-1;
    border-radius: 8px;
    border: 1px solid $white-4;
    padding: 2.8vw;
    margin-top: 4.5vh;
    margin-bottom: 4.5vh;

    & .calculatorHeader {
      display: flex;
      flex-direction: column;
      align-content: flex-start;
      justify-content: center;

      & img {
        align-self: flex-start;
      }

      & h2 {
        font-weight: 600;
        font-size: calc(#{$font} * 2);
        line-height: 40px;
      }

      & h4 {
        font-weight: normal;
        font-size: calc(#{$font});
        color: $gray-1;
        line-height: 20px;
      }
    }

    & .calculatorInputs {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 17px;
      grid-auto-flow: column;
      width: 100%;
      margin-top: 40px;

      & label {
        font-weight: 500;
        font-size: $font;
        line-height: 20px;
      }
    }

    & .summary {
      display: flex;
      margin-top: 25px;
      flex-direction: column;
      border: 1px solid $white-4;
      box-shadow: 0px 1px 4px rgba(150, 164, 176, 0.1);
      border-radius: 4px;

      & .depositSummary {
        display: flex;
        align-content: center;
        justify-content: space-between;
        padding: 30px;

        & h3 {
          font-weight: 500;
          font-size: calc(#{$font} * 1.1);
          color: $black-1;
          align-self: center;
          padding: 0;

          & .amountText {
            display: inline-block;
          }
        }

        & h2 {
          font-weight: 500;
          font-size: 40px;
          color: $blue-2;
          text-align: center;
        }
      }

      & .descriptionSummary {
        display: flex;
        align-content: center;
        justify-content: space-between;
        background-color: $white-2;
        padding: 15px 30px;

        & p {
          color: $black-1;
          font-weight: normal;
          font-size: $font;
          line-height: 22px;
        }
      }
    }

    & .confirmButton {
      display: flex;
      align-content: center;
      justify-content: center;
      padding: 30px 30px 0 30px;

      & button {
        width: 100%;
        height: 56px;
        background-color: $blue-1;
        border-radius: 32px;

        font-weight: 600;
        font-size: 18px;
        line-height: 24px;
        color: $white-1;

        &:disabled {
          background: $blue-1;
          opacity: 0.2;
        }
      }
    }
  }
}

//PHONE RESOLUTION//

@media all and (max-width: 768px) {
  .savingGoalSimulator {
    max-width: 100%;
    min-width: 100%;

    & .goalCalculator {
      background-color: $white-1;
      border: 0;
      border-radius: 12px;
      padding: 25px;
      margin-top: 3.2vh;
      margin-bottom: 3.2vh;

      & .calculatorHeader {
        & img {
          align-self: flex-start;
        }

        & h2 {
          font-size: calc(#{$font} * 1.8);
        }

        & h4 {
          font-size: calc(#{$font} * 1);
        }
      }

      & .calculatorInputs {
        display: flex;
        flex-direction: column;
        width: 100%;
        margin-top: 30px;

        & label {
          font-size: calc(#{$font} * 1.6);
        }
      }

      & .summary {
        & .depositSummary {
          & h3 {
            font-size: 16px;

            & .amountText {
              display: none;
            }
          }

          & h2 {
            font-size: 26px;
          }
        }
      }

      & .confirmButton {
        padding: 30px 0px;
      }
    }
  }
}
