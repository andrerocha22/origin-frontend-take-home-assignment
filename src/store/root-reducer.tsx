import { Action, Reducer, Dispatch } from 'redux';

export interface InitialState {
  year: number;
  month: string;
  amount: number;
}

export const initialState: InitialState = {
  year: 1999,
  month: 'January',
  amount: null
};

export interface DispatchAction extends Action {
  payload: Partial<InitialState>;
}

export enum ActionType {
  UpdateYear,
  UpdateMonth,
  UpdateAmount
}

export const rootReducer: Reducer<InitialState, DispatchAction> = (
  state = initialState,
  action
) => {
  //Trocar para switch
  if (action.type === ActionType.UpdateYear) {
    return { ...state, year: action.payload.year || 0 };
  } else if (action.type === ActionType.UpdateMonth) {
    return { ...state, month: action.payload.month || '' };
  } else if (action.type === ActionType.UpdateAmount) {
    return { ...state, amount: action.payload.amount || 0 };
  } else return state;
};

export class RootDispatcher {
  dispatch: Dispatch<DispatchAction>;

  constructor(dispatch: Dispatch<DispatchAction>) {
    this.dispatch = dispatch;
  }

  updateYear = (year: number) =>
    this.dispatch({ type: ActionType.UpdateYear, payload: { year } });

  updateMonth = (month: string) =>
    this.dispatch({ type: ActionType.UpdateMonth, payload: { month } });

  updateAmount = (amount: number) =>
    this.dispatch({ type: ActionType.UpdateAmount, payload: { amount } });
}
