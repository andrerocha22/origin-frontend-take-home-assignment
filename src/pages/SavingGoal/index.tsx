import * as React from 'react';
import './styles.scss';
import ValueInput from '../../components/ValueInput';
import DateInput from '../../components/DateInput';

import { useSelector } from 'react-redux';
import { InitialState } from '../../store/root-reducer';

interface StateProps {
  year: number;
  month: string;
  amount: number;
  deposit: number;
}

const SavingGoal: React.FC = () => {
  const { year, month, amount, deposit } = useSelector<
    InitialState,
    StateProps
  >((state: InitialState) => {
    return {
      year: state.year,
      month: state.month,
      amount: state.amount,
      deposit: state.deposit
    };
  });

  const depositValue = () => {
    const valueToString = amount.toString(10).replace(/,/g, '');
    const depositValue = Number(valueToString) / deposit;
    return depositValue.toFixed(2);
  };

  return (
    <section className="savingGoalSimulator">
      <h3>
        Let&apos;s plan your <strong>saving goal.</strong>
      </h3>
      <div className="goalCalculator">
        <div className="calculatorHeader">
          <img src={require('~/assets/icons/house.svg')} alt="house-icon" />
          <h2>Buy a house</h2>
          <h4>Saving goal</h4>
        </div>
        <div className="calculatorInputs">
          <ValueInput />
          <DateInput />
        </div>

        <div className="summary">
          <span className="depositSummary">
            <h3>
              Monthly <span className="amountText">amount</span>
            </h3>
            <h2 className="depositValue">${depositValue()}</h2>
          </span>
          <span className="descriptionSummary">
            <p>
              You are plaining{' '}
              <strong>
                {deposit} monthly {deposit > 1 ? 'deposits' : 'deposit'}
              </strong>{' '}
              to reach your <strong>${amount} </strong>goal by{' '}
              <strong>
                {month} {year}
              </strong>
              .
            </p>
          </span>
        </div>
        <div className="confirmButtonContainer">
          <button className='confirmButton' disabled={amount === 0 ? true : false}>Confirm</button>
        </div>
      </div>
    </section>
  );
};

export default SavingGoal;
