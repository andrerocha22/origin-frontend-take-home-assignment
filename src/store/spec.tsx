import { RootDispatcher } from './root-reducer';

import configureStore from 'redux-mock-store';
const mockStore = configureStore();
const store = mockStore();

export enum ActionType {
  UpdateYear,
  UpdateMonth,
  UpdateAmount,
  UpdateDeposit
}

describe('input_actions', () => {
  const rootDispatcher = new RootDispatcher(store.dispatch);

  beforeEach(() => {
    store.clearActions();
  });
  test('Dispatches the correct action and payload for year', () => {
    const expectedActions = [
      {
        payload: { year: 1999 },
        type: ActionType.UpdateYear
      }
    ];

    rootDispatcher.updateYear(1999);
    expect(store.getActions()).toEqual(expectedActions);
  });

  test('Dispatches the correct action and payload for month', () => {
    const expectedActions = [
      {
        payload: { month: 'January' },
        type: ActionType.UpdateMonth
      }
    ];

    rootDispatcher.updateMonth('January');
    expect(store.getActions()).toEqual(expectedActions);
  });

  test('Dispatches the correct action and payload for amount', () => {
    const expectedActions = [
      {
        payload: { amount: 459.99 },
        type: ActionType.UpdateAmount
      }
    ];

    rootDispatcher.updateAmount(459.99);
    expect(store.getActions()).toEqual(expectedActions);
  });

  test('Dispatches the correct action and payload for deposit', () => {
    const expectedActions = [
      {
        payload: { deposit: 12 },
        type: ActionType.UpdateDeposit
      }
    ];

    rootDispatcher.updateDeposit(12);
    expect(store.getActions()).toEqual(expectedActions);
  });
});
