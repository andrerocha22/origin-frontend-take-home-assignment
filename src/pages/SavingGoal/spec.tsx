import React from 'react';
import { mount } from 'enzyme';

import SavingGoal from './index';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';

import { createStore } from 'redux';

import { rootReducer, RootDispatcher } from '../../store/root-reducer';

describe('<SavingGoal/>', () => {
  const initialState = {
    year: 1999,
    month: 'January',
    amount: 0.0,
    deposit: 1
  };

  let store, wrapper, rootDispatcher;

  beforeEach(() => {
    store = createStore(rootReducer, initialState);
    wrapper = mount(
      <Provider store={store}>
        <SavingGoal />
      </Provider>
    );
    rootDispatcher = new RootDispatcher(store.dispatch);
  });

  it('Renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render without errors', () => {
    expect(wrapper.find('.savingGoalSimulator').length).toEqual(1);
  });

  it('Should return the correct format deposit value (5000 -> 5000/1 month -> $5000.00)', () => {
    rootDispatcher.updateAmount(5000);
    expect(wrapper.find('.depositValue').text()).toEqual('$5000.00');
  });

  it('Should return the correct format deposit value (5000 -> 5000/5 months -> $1000.00)', () => {
    rootDispatcher.updateAmount(5000);

    for (let i = 0; i < 4; i++) {
      wrapper.find('.forward').simulate('click');
    }

    expect(wrapper.find('.depositValue').text()).toEqual('$1000.00');
    expect(wrapper.find('p').text()).toEqual(
      'You are plaining 5 monthly deposits to reach your $5000 goal by September 2020.'
    );
  });

  it('Should return the correct format deposit value (5000 -> 5000/4 months -> $1250.00)', () => {
    rootDispatcher.updateAmount(5000);

    for (let i = 0; i < 3; i++) {
      wrapper.find('.forward').simulate('click');
    }
    wrapper.find('.backward').simulate('click');
    wrapper.find('.forward').simulate('click');

    expect(wrapper.find('.depositValue').text()).toEqual('$1250.00');
    expect(wrapper.find('p').text()).toEqual(
      'You are plaining 4 monthly deposits to reach your $5000 goal by August 2020.'
    );
  });

  it('Should enable the button when the value is changed', () => {
    const input = wrapper.find('input');
    input.instance().value = '500000';
    input.simulate('change');

    const button = wrapper.find('.confirmButton');
    expect(button.props()['disabled']).toBe(false);
  });
});
