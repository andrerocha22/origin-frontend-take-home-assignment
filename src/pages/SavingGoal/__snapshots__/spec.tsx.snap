// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SavingGoal/> Renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <SavingGoal>
    <section
      className="savingGoalSimulator"
    >
      <h3>
        Let's plan your 
        <strong>
          saving goal.
        </strong>
      </h3>
      <div
        className="goalCalculator"
      >
        <div
          className="calculatorHeader"
        >
          <img
            alt="house-icon"
            src="test-file-stub"
          />
          <h2>
            Buy a house
          </h2>
          <h4>
            Saving goal
          </h4>
        </div>
        <div
          className="calculatorInputs"
        >
          <ValueInput>
            <div
              className="value"
            >
              <label>
                Total amount
              </label>
              <span
                className="valueInput"
              >
                <span
                  className="currencyIcon"
                >
                  $
                </span>
                <CurrencyInput
                  allowNegative={false}
                  autoFocus={true}
                  className="inputAmount"
                  decimalSeparator="."
                  inputType="text"
                  maxLength="12"
                  onChange={[Function]}
                  onChangeEvent={[Function]}
                  precision="2"
                  prefix=""
                  selectAllOnFocus={true}
                  suffix=""
                  thousandSeparator=","
                  value={0}
                >
                  <input
                    className="inputAmount"
                    maxLength="12"
                    onChange={[Function]}
                    onFocus={[Function]}
                    onMouseUp={[Function]}
                    type="text"
                    value="0.00"
                  />
                </CurrencyInput>
              </span>
            </div>
          </ValueInput>
          <DateInput>
            <div
              className="date"
              data-test="setDateComponent"
            >
              <label>
                Reach goal by
              </label>
              <div
                className="dateInput"
              >
                <button
                  className="backward"
                  disabled={true}
                  onClick={[Function]}
                  onKeyUp={[Function]}
                >
                  <img
                    alt="arrow-left"
                    src="test-file-stub"
                  />
                </button>
                <span
                  className="dateDisplay"
                >
                  <h1
                    className="month"
                  >
                    May
                  </h1>
                  <h2
                    className="year"
                  >
                    2020
                  </h2>
                </span>
                <button
                  className="forward"
                  onClick={[Function]}
                  onKeyUp={[Function]}
                >
                  <img
                    alt="arrow-right"
                    src="test-file-stub"
                  />
                </button>
              </div>
            </div>
          </DateInput>
        </div>
        <div
          className="summary"
        >
          <span
            className="depositSummary"
          >
            <h3>
              Monthly 
              <span
                className="amountText"
              >
                amount
              </span>
            </h3>
            <h2
              className="depositValue"
            >
              $
              0.00
            </h2>
          </span>
          <span
            className="descriptionSummary"
          >
            <p>
              You are plaining
               
              <strong>
                1
                 monthly 
                deposit
              </strong>
               
              to reach your 
              <strong>
                $
                0
                 
              </strong>
              goal by
               
              <strong>
                May
                 
                2020
              </strong>
              .
            </p>
          </span>
        </div>
        <div
          className="confirmButtonContainer"
        >
          <button
            className="confirmButton"
            disabled={true}
          >
            Confirm
          </button>
        </div>
      </div>
    </section>
  </SavingGoal>
</Provider>
`;
